contracts/ERC1155NftCustom.sol

```diff
@@ -0,0 +1,155 @@
+        baseURI = _initBaseURI;
+        _owner = owner;
+        name = _name;
+        symbol = _symbol;
+    }
+
+    function setURI(string memory _newURI) public onlyRole(MINTER_ROLE) {
+        _setURI(_newURI);
+    }
+
+    function updateTokenUri(uint256 _tokenId, string memory _newUri, bool _isFreezeTokenUri)
+    public
+    onlyRole(MINTER_ROLE) {
+        require(_exists(_tokenId), "NFT: update URI query for nonexistent token");
+        require(isFreezeTokenUris == false, "NFT: Token uris are frozen globally");
+        require(freezeTokenUris[_tokenId] != true, "NFT: Token is frozen");
+        require(_isFreezeTokenUri || (bytes(_newUri).length != 0), "NFT: Either _newUri or _isFreezeTokenUri=true required");
+
+        if (bytes(_newUri).length != 0) {
+            require(keccak256(bytes(_tokenURIs[_tokenId])) != keccak256(bytes(string(abi.encodePacked(_newUri)))), "NFT: New token URI is same as updated");
+            _tokenURIs[_tokenId] = _newUri;
+            emit URI(_newUri, _tokenId);
+        }
+        if (_isFreezeTokenUri) {
+            freezeTokenUris[_tokenId] = true;
+            emit PermanentURI(_tokenURIs[_tokenId], _tokenId);
+        }
+    }
+
+
+    function update(string memory _newBaseURI, bool _freezeAllTokenUris)
+    public
+    onlyRole(MINTER_ROLE) {
+        require(isFreezeTokenUris == false, "NFT: Token uris are already frozen");
+        baseURI = _newBaseURI;
+        if (_freezeAllTokenUris) {
+            freezeAllTokenUris();
+        }
+    }
+
+    function freezeAllTokenUris()
+    public
+    onlyRole(MINTER_ROLE) {
+        require(isFreezeTokenUris == false, "NFT: Token uris are already frozen");
+        isFreezeTokenUris = true;
+
+        emit PermanentURIGlobal();
+    }
+
+    function totalSupply (uint256 _id) public view returns (uint256) {
+        return tokenSupply[_id];
+    }
+
+    function uri(uint256 _id) public override view returns (string memory) {
+        if (bytes(_tokenURIs[_id]).length > 0) {
+            if (bytes(baseURI).length > 0) {
+                return string(abi.encodePacked(baseURI, _tokenURIs[_id]));
+            } else {
+                return _tokenURIs[_id];
+            }
+        } else {
+            return super.uri(_id);
+        }
+    }
+
+    function mintToCaller( address account, uint256 id, uint256 amount, string memory uri) public onlyRole(MINTER_ROLE) {
+        require(!_exists(id), "NFT: token already minted");
+        if (bytes(uri).length > 0) {
+            _tokenURIs[id] = uri;
+            emit URI(uri, id);
+        }
+        _mint(account, id, amount, "");
+        tokenSupply[id] += amount;
+    }
+
+    function mintToCallerBatch(
+        address to,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        string[] memory uris
+    ) public onlyRole(MINTER_ROLE) {
+        for (uint256 i = 0; i < ids.length; i++) {
+            require(!_exists(ids[i]), "NFT: one of tokens are already minted");
```

Comment:
should this check be run for all ids before any values are changed? Since atm it seems like tokenSupply and _tokenURIs arrays might get updated, URI event emitted, but minting might not actually happen

Also should we check that the "amounts" array is as long as ids? And uris? (check ERC1155.sol _mintBatch function)

Reply:
Will check it

Reply:
Regarding pre-require actions, if revert is fired, state will not be updated.
Refer to
https://ethereumdev.io/handle-errors-in-solidity-with-require-and-revert/
> Solidity manages errors by using state-reverting exceptions. These exceptions revert all modifications done to the state in the current call, including all sub-calls. Additionally, it can show an error to the caller if a string is passed as parameter.

Lengths of arrays check will be added

Reply:
Added check for non-zero amounts value, added corresponding tests